File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/core/usecases/patrimonies/CreatePatrimony.kt
package dev.auguste.agni_api.core.usecases.patrimonies

import dev.auguste.agni_api.core.adapters.repositories.IRepository
import dev.auguste.agni_api.core.adapters.repositories.IUnitOfWork
import dev.auguste.agni_api.core.entities.Account
import dev.auguste.agni_api.core.entities.Patrimony
import dev.auguste.agni_api.core.entities.PatrimonySnapshot
import dev.auguste.agni_api.core.entities.enums.PatrimonySnapshotStatusType
import dev.auguste.agni_api.core.usecases.CreatedOutput
import dev.auguste.agni_api.core.entities.DomainException
import dev.auguste.agni_api.core.usecases.interfaces.IUseCase
import dev.auguste.agni_api.core.usecases.patrimonies.dto.CreatePatrimonyInput
import java.time.LocalDate
import java.util.Date

class CreatePatrimony(
    private val patrimonyRepo: IRepository<Patrimony>,
    private val accountRepo: IRepository<Account>,
    private val snapshotRepo: IRepository<PatrimonySnapshot>,
    private val unitOfWork: IUnitOfWork): IUseCase<CreatePatrimonyInput, CreatedOutput> {

    override fun execAsync(input: CreatePatrimonyInput): CreatedOutput {
        return unitOfWork.execute {
            if (patrimonyRepo.existsByName(input.title))
                throw DomainException.AlreadyExist.Patrimony(input.title)

            if (input.accountIds.isNotEmpty())
                if (this.accountRepo.getManyByIds(input.accountIds).size != input.accountIds.size)
                    throw DomainException.NotFound.SomeAccounts(input.accountIds)

            val newPatrimony = Patrimony(
                title = input.title,
                amount = input.amount,
                accountIds = input.accountIds.toMutableSet(),
                type = input.type
            )

            patrimonyRepo.create(newPatrimony)

            if (newPatrimony.amount > 0) {
                val firstSnapShot = PatrimonySnapshot(
                    patrimonyId = newPatrimony.id,
                    currentBalanceObserved = newPatrimony.amount,
                    date = LocalDate.now(),
                    status = PatrimonySnapshotStatusType.COMPLETED
                )

                snapshotRepo.create(firstSnapShot)
            }

            CreatedOutput(newPatrimony.id)
        }
    }
}

File: apps/agni-server_archived/src/core/interactions/patrimony/createPatrimony.ts
import { CreatedDto } from "@core/dto/base";
import { IUsecase } from "../interfaces";
import { ResourceAlreadyExist } from "@core/errors/resourceAlreadyExistError";
import { ResourceNotFoundError } from "@core/errors/resournceNotFoundError";
import { Patrimony } from "@core/domains/entities/patrimony";
import { GetUID } from "@core/adapters/libs";
import { mapperPatrimonyType, TransactionStatus } from "@core/domains/constants";
import { UnitOfWorkRepository } from "@core/repositories/unitOfWorkRepository";
import { PatrimonySnapshot } from "@core/domains/entities/patrimonySnapshot";
import Repository from "@core/adapters/repository";
import { Account } from "@core/domains/entities/account";

export type RequestCreatePatrimony = {
    title: string
    amount: number
    accountIds?: string[]
    type: string
}

export class CreatePatrimonyUseCase implements IUsecase<RequestCreatePatrimony, CreatedDto> {
    private patrimonyRepo: Repository<Patrimony>
    private accountRepo: Repository<Account>
    private snapshotRepo: Repository<PatrimonySnapshot>
    private unitOfWork: UnitOfWorkRepository

    constructor(
        patrimonyRepo: Repository<Patrimony>, 
        accountRepo: Repository<Account>,
        snapshotRepo: Repository<PatrimonySnapshot>,
        unitOfWork: UnitOfWorkRepository
    ) {
        this.accountRepo = accountRepo
        this.patrimonyRepo = patrimonyRepo
        this.snapshotRepo = snapshotRepo
        this.unitOfWork = unitOfWork
    }

    async execute(request: RequestCreatePatrimony): Promise<CreatedDto> {
        try {
            const trx = await this.unitOfWork.start()

            if (await this.patrimonyRepo.existByName(request.title))
                throw new ResourceAlreadyExist("PATRIMONY_ALREADY_EXIST")

            if (request.accountIds && request.accountIds.length > 1) {
                if ((await this.accountRepo.getManyByIds(request.accountIds)).length === 0)
                    throw new ResourceNotFoundError("SOME_ACCOUNT_NOT_FOUND")
            }

            const patrimony = new Patrimony(GetUID(), request.title, request.amount, 
                mapperPatrimonyType(request.type), request.accountIds ? request.accountIds : []) 

            await this.patrimonyRepo.create(patrimony, trx)

            if (patrimony.getAmount() > 0) {
                const firstSnapshot = new PatrimonySnapshot(GetUID(), 
                    patrimony.getId(), patrimony.getAmount(), TransactionStatus.COMPLETE, 
                    new Date()
                );
                await this.snapshotRepo.create(firstSnapshot, trx);
            } 

            await this.unitOfWork.commit()

            return { newId: patrimony.getId() }
            
        } catch(err) {
            await this.unitOfWork.rollback()
            throw err
        } 
    }
}